### ISSUE TITLE
CTCLoss produces NaNs in some situations

### ISSUE THREAD

**Issue by WenmuZhou** [pytorch/pytorch#14335]
## 🐛 Bug

when I train a cnn-rnn-ctc text recognize model, I meet nan loss after some iters, but it's ok at pytorch 0.4 with warpctc

## To Reproduce

1. download the code from https://github.com/WenmuZhou/crnn.pytorch
2. change the ctc loss from warpctc to nn.CTCloss()
3. run

 - PyTorch Version (e.g., 1.0):  1.0.0.dev20181115
 - OS (e.g., Linux): ubuntu 16.04
 - How you installed PyTorch (`conda`, `pip`, source): pip3
 - Build command you used (if compiling from source):
 - Python version: 3.5.2
 - CUDA/cuDNN version: 8.0/6.0
 - GPU models and configuration: 1080ti
 - Any other relevant information:

**Comment by soumith** [comment]
warpctc and nn.CTCLoss have slightly different API. Consult the docs for more information https://pytorch.org/docs/master/nn.html#torch.nn.CTCLoss

**Comment by WenmuZhou** [comment]
@soumith Accroding to the doc, I have add a log_softmax to the output from network ,but the nan is appeared in later iters.

**Comment by viig99** [comment]
Same issue, nan pops up, after running through log_softmax

**Comment by gaochangw** [comment]
I experience the same problem when training an RNN. Getting nan after a few epochs if I replace the CTCLoss in warpctc_pytorch with the CTCLoss in torch.nn.

**Comment by ezyang** [comment]
I believe #21244 will bandaid fix this problem.

**Comment by yqwangustc** [comment]
https://github.com/pytorch/pytorch/pull/21392 is landed. Please let me know if you still have NaN.

**Comment by fmassa** [comment]
Closing via #21392

**Comment by Marchbruno09** [comment]
Or try to reduce learning rate to 0.001

**Comment by ppriyank** [comment]
**For the NaN in gradients, this happens when the loss is infinite.** this is not necessarily the case 
I do this : 
`loss = criterion(softmax(float_out), targets, output_sizes, target_sizes)`
I get the loss to be : `tensor(1.2917, device='cuda:0', grad_fn=<MeanBackward0>)` and 
and gradient is `tensor([[[[nan, nan, nan,  ...,]]]], device='cuda:0')`

but if I do 

`loss = criterion(softmax(float_out).cpu(), targets, output_sizes, target_sizes) #tensor(1.2917, grad_fn=<MeanBackward0>)`

I get the right non-nan gradients.

**Comment by t-vi** [comment]
Do you have the inputs?

**Comment by ppriyank** [comment]
yes

**Comment by t-vi** [comment]
Awesome, thank you! The mail address on my github page would work well.
